vignettes/articles/ui-cli-conversion.Rmd

```diff
@@ -0,0 +1,442 @@
+
+```{r}
+#| include: false
+ui_code_snippet <- usethis:::ui_code_snippet
+usethis_theme <- usethis:::usethis_theme
+ui_inform <- usethis:::ui_inform
+kv_line <- usethis:::kv_line
+ui_special <- usethis:::ui_special
+ui_abort <- usethis:::ui_abort
+ui_bullets <- usethis:::ui_bullets
+```
+
+## Block styles
+
+The block styles exist to produce bulleted output with a specific symbol, using a specific color.
+
+```{asciicast}
+#| collapse: false
+f <- function() {
+  ui_todo("ui_todo(): red bullet")
+  ui_done("ui_done(): green check")
+  ui_oops("ui_oops(): red x")
+  ui_info("ui_info(): yellow i")
+  ui_line("ui_line(): (no symbol)")
+}
+f()
+```
+
+Another important feature is that all of this output can be turned off package-wide via the `usethis.quiet` option.
+
+```{asciicast}
+withr::with_options(
+  list(usethis.quiet = TRUE),
+  ui_info("You won't see this message.")
+)
+withr::with_options(
+  list(usethis.quiet = FALSE), # this is the default
+  ui_info("But you will see this one.")
+)
+```
+
+These styles are very close to what can be done with `cli::cli_bullets()` and the way it responds to the names of its input `text`.
+
+```{asciicast}
+cli::cli_bullets(c(
+        "noindent",
+  " " = "indent",
+  "*" = "bullet",
+  ">" = "arrow",
+  "v" = "success",
+  "x" = "danger",
+  "!" = "warning",
+  "i" = "info"
+))
+```
+
+A direct translation would look something like this:
+
+| Legacy `ui_*()` | `cli_bullets()` shortcode | tweaks needed                 |
+|-----------------|---------------------------|-------------------------------|
+| `ui_todo()`     | `*`                       | blue (default) -\> red        |
+| `ui_done()`     | `v`                       | perfect match                 |
+| `ui_oops()`     | `x`                       | perfect match                 |
+| `ui_info()`     | `i`                       | blue (default) -\> yellow     |
+| `ui_line()`     | (unnamed)                 | sort of a perfect match? although sometimes `ui_line()` is used just to get a blank line |
+
+The overall conversion plan is to switch to a new function, `ui_bullets()`, which is a wrapper around `cli::cli_bullets()`, that adds a few features:
+
+* Early exit, without emitting messages, if the `usethis.quiet` option is `TRUE`.
+* A usethis theme, that changes the color of certain bullet styles and adds a new style for todo's.
+
+```{asciicast}
+ui_bullets(c(
+  "v" = "A great success!",
+  "_" = "Something you need to do.",
+  "x" = "Bad news.",
+  "i" = "The more you know.",
+  " " = "I'm just here for the indentation.",
+  "No indentation at all. Not used much in usethis."
+))
+```
+
+
+Summary of what I've done for todo's:
+
+-   Introduce a new bullet shortcode for a todo. Proposal: `_` (instead of `*`), which seems to be the best single ascii character that evokes a place to check something off.
```

Comment:
`[]` would another option. It's not a single character but it does better evoke the checkbox.

Reply:
I think I care more about the shortcode being single character and, therefore, lining up nicely with its friends in code, than about evoking the checkbox.